src/Compilers/Test/Utilities/CSharp/CompilationTestUtils.cs

```diff
@@ -278,5 +278,107 @@ public static List<string> LookupNames(this SemanticModel model, int position, I
                     : model.LookupSymbols(position, container);
             return symbols.Select(s => s.Name).Distinct().ToList();
         }
+
+        /// <summary>
+        /// Verify the type and nullability inferred by NullabilityWalker of all expressions in the source
+        /// that are followed by specific annotations. Annotations are of the form /*T:type*/.
+        /// </summary>
+        internal static void VerifyTypes(this CSharpCompilation compilation, SyntaxTree tree = null)
+        {
+            if (tree == null)
+            {
+                foreach (var syntaxTree in compilation.SyntaxTrees)
+                {
+                    VerifyTypes(compilation, syntaxTree);
+                }
+                return;
+            }
+
+            var root = tree.GetRoot();
+            var allAnnotations = getAnnotations();
+            if (allAnnotations.IsEmpty)
+            {
+                return;
+            }
+
+            var model = compilation.GetSemanticModel(tree);
+            var annotationsByMethod = allAnnotations.GroupBy(annotation => annotation.Expression.Ancestors().OfType<BaseMethodDeclarationSyntax>().First()).ToArray();
+            foreach (var annotations in annotationsByMethod)
+            {
+                var method = (MethodSymbol)model.GetDeclaredSymbol(annotations.Key);
+                var diagnostics = DiagnosticBag.GetInstance();
+                var block = MethodCompiler.BindMethodBody(method, new TypeCompilationState(method.ContainingType, compilation, null), diagnostics);
+                var dictionary = new Dictionary<SyntaxNode, TypeSymbolWithAnnotations>();
+                NullableWalker.Analyze(
+                    compilation,
+                    method,
+                    block,
+                    diagnostics,
+                    (BoundExpression expr, TypeSymbolWithAnnotations exprType) => dictionary[expr.Syntax] = exprType);
+                diagnostics.Free();
+                var expectedTypes = annotations.SelectAsArray(annotation => annotation.Text);
+                var actualTypes = annotations.SelectAsArray(annotation => toDisplayString(annotation.Expression));
+                // Consider reporting the correct source with annotations on mismatch.
+                AssertEx.Equal(expectedTypes, actualTypes);
+
+                string toDisplayString(SyntaxNode syntaxOpt)
```

Comment:
Nit: I'd move this local one line down (outside this block)

Reply:
Unfortunately, that would require passing `dictionary` as a parameter. Less (code) is more.